### PR TITLE
Prove that `dupIO` can solve the `Source` problem

### PR DESCRIPTION
**Pull request by edsko**

### CODE REVIEW

File: cbits/dup-prim.cmm [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        }
+    }
+
+    type_not_ok:
+    foreign "C" dupUnsupportedWarning(clos "ptr");
+    return (clos);
+
+    type_ok:
+    W_ len;
+    (len) = foreign "C" dupClosureSize(clos "ptr");
```
Comment (AndreasPK) [inline cbits/dup-prim.cmm]: Why not call closure_sizeW directly here?

File: cbits/dup-prim.cmm [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    W_ info, type;
+    clos = UNTAG(R1);
+    info = %INFO_PTR(clos);
+    type = TO_W_(%INFO_TYPE(%STD_INFO(info)));
+    switch [0 .. N_CLOSURE_TYPES] type {
+        case FUN, FUN_1_0, FUN_0_1, FUN_1_1, FUN_2_0, FUN_0_2, FUN_STATIC: {
+            goto type_ok;
+        }
+        // Do not copy data without pointers
+        // (includes static data such as)
```
Comment (AndreasPK) [inline cbits/dup-prim.cmm]: such as what? I guess it refers to the constructor list below?

File: cbits/dup-prim.cmm (hunk 1 above)
Comment (edsko) [reply cbits/dup-prim.cmm]: No idea :) Will fix.

File: src/Data/Dup/Internal.hs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import GHC.Prim
+import GHC.Exts
+
+foreign import prim "dupClosure" dupClosure# ::
+  forall s. Any -> State# s -> (# State# s, Any #)
+
+{-# NOINLINE dup# #-}
+dup# :: forall a s. a -> State# s -> (# State# s, a #)
+dup# a s0 =
+    case dupClosure# (unsafeCoerce# a) s0
```
Comment (AndreasPK) [inline src/Data/Dup/Internal.hs:19]: (unsafeCoerce# dupClosure#) seems simpler.

Comment (edsko) [reply src/Data/Dup/Internal.hs:19]: Would it not be better to keep the scope of `unsafeCoerce(#)` as small as possible?

**Pull request merged**